File: frontend/lib/features/planner/models/planner_event.dart
import 'event_status.dart';

/// A simple time-blocked planner event, mirroring {@code EventResponse}.
class PlannerEvent {
  PlannerEvent({
    required this.id,
    required this.title,
    required this.status,
    required this.startAt,
    required this.endAt,
    this.description,
    this.location,
    this.createdAt,
    this.updatedAt,
  });

  final String id;
  final String title;
  final String? description;
  final String? location;
  final DateTime startAt;
  final DateTime endAt;
  final EventStatus status;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  factory PlannerEvent.fromJson(Map<String, dynamic> json) {
    final startRaw = json['startAt'];
    final endRaw = json['endAt'];
    final createdAtRaw = json['createdAt'];
    final updatedAtRaw = json['updatedAt'];
    return PlannerEvent(
      id: json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      location: json['location'] as String?,
      startAt: DateTime.parse(startRaw as String),
      endAt: DateTime.parse(endRaw as String),
      status: EventStatus.fromWire(json['status'] as String?),
      createdAt: createdAtRaw is String ? DateTime.parse(createdAtRaw) : null,
      updatedAt: updatedAtRaw is String ? DateTime.parse(updatedAtRaw) : null,
    );
  }
}

File: backend/src/main/java/com/blistra/planner/domain/PlannerEvent.java
package com.blistra.planner.domain;

import com.blistra.users.domain.User;
import jakarta.persistence.*;

import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.util.UUID;

/**
 * A generic personal planning event (meeting, personal appointment, study
 * session, travel, ...). Planner events are owned by the Planner module and
 * never duplicate records owned by other modules (health appointments,
 * medication schedules, habit schedules).
 */
@Entity
@Table(name = "planner_event",
        indexes = {
                @Index(name = "idx_planner_event_user_start", columnList = "user_id, start_at"),
                @Index(name = "idx_planner_event_user_status", columnList = "user_id, status")
        })
public class PlannerEvent {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "UUID")
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false, updatable = false)
    private User user;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(length = 2000)
    private String description;

    @Column(length = 255)
    private String location;

    @Column(name = "start_at", nullable = false)
    private OffsetDateTime startAt;

    @Column(name = "end_at", nullable = false)
    private OffsetDateTime endAt;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private EventStatus status;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    public PlannerEvent() {
        this.status = EventStatus.SCHEDULED;
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
    }

    public OffsetDateTime getStartAt() {
        return startAt;
    }

    public void setStartAt(OffsetDateTime startAt) {
        this.startAt = startAt;
    }

    public OffsetDateTime getEndAt() {
        return endAt;
    }

    public void setEndAt(OffsetDateTime endAt) {
        this.endAt = endAt;
    }

    public EventStatus getStatus() {
        return status;
    }

    public void setStatus(EventStatus status) {
        this.status = status;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}